packages/flutter/lib/src/rendering/object.dart

```diff
@@ -671,6 +672,105 @@ class _ForkingSemanticsFragment extends _SemanticsFragment {
   }
 }
 
+class RenderObjectOwner {
+
+  List<RenderObject> _nodesNeedingLayout = <RenderObject>[];
+  bool _debugDoingLayout = false;
+  bool get debugDoingLayout => _debugDoingLayout;
+  /// Update the layout information for all dirty render objects.
+  ///
+  /// This function is one of the core stages of the rendering pipeline. Layout
+  /// information is cleaned prior to painting so that render objects will
+  /// appear on screen in their up-to-date locations.
+  ///
+  /// See [FlutterBinding] for an example of how this function is used.
+  void flushLayout() {
+    Timeline.startSync('Layout');
+    _debugDoingLayout = true;
+    try {
+      // TODO(ianh): assert that we're not allowing previously dirty nodes to redirty themeselves
+      while (_nodesNeedingLayout.isNotEmpty) {
+        List<RenderObject> dirtyNodes = _nodesNeedingLayout;
+        _nodesNeedingLayout = <RenderObject>[];
+        for (RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => a.depth - b.depth)) {
+          if (node._needsLayout && node.attached)
+            node._layoutWithoutResize();
+        }
+      }
+    } finally {
+      _debugDoingLayout = false;
+      Timeline.finishSync();
+    }
+  }
+
+  List<RenderObject> _nodesNeedingCompositingBitsUpdate = <RenderObject>[];
+  /// Updates the [needsCompositing] bits.
+  ///
+  /// Called as part of the rendering pipeline after [flushLayout] and before
+  /// [flushPaint].
+  void flushCompositingBits() {
+    Timeline.startSync('Compositing Bits');
+    _nodesNeedingCompositingBitsUpdate.sort((RenderObject a, RenderObject b) => a.depth - b.depth);
+    for (RenderObject node in _nodesNeedingCompositingBitsUpdate) {
+      if (node.attached)
+        node._updateCompositingBits();
+    }
+    _nodesNeedingCompositingBitsUpdate.clear();
+    Timeline.finishSync();
+  }
+
+  List<RenderObject> _nodesNeedingPaint = <RenderObject>[];
+  bool _debugDoingPaint = false;
+  bool get debugDoingPaint => _debugDoingPaint;
+  /// Update the display lists for all render objects.
+  ///
+  /// This function is one of the core stages of the rendering pipeline.
+  /// Painting occurs after layout and before the scene is recomposited so that
+  /// scene is composited with up-to-date display lists for every render object.
+  ///
+  /// See [FlutterBinding] for an example of how this function is used.
+  void flushPaint() {
+    Timeline.startSync('Paint');
+    _debugDoingPaint = true;
+    try {
+      List<RenderObject> dirtyNodes = _nodesNeedingPaint;
+      _nodesNeedingPaint = <RenderObject>[];
+      // Sort the dirty nodes in reverse order (deepest first).
+      for (RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => b.depth - a.depth)) {
+        assert(node._needsPaint);
+        if (node.attached)
```

Comment:
We should probably assert that if owner != null, then owner == this. Otherwise we screwed up the ownership chain.